aten/src/ATen/native/Histogram.cpp

```diff
@@ -6,117 +6,181 @@
+ *               for each dimension.
+ *   range     - float[] of length D, optional. If specified, defines the leftmost and rightmost bin edges
+ *               for each dimension.
+ *   weight    - tensor, optional. If provided, weight should have the same shape as input excluding its last dimension.
+ *               Each D-dimensional coordinate in input contributes its associated weight towards its bin's result.
+ *               If weight is not specified, each coordinate has weight 1 by default.
  *   density   - bool, optional. If False, the result will contain the number of samples (or total weight)
  *               in each bin. If True, the result is the value of the probability density function at the
- *               bin, normalized such that the integral over the range is 1.
+ *               bin, normalized such that the integral over all bins is 1.
  *
  * Returns:
- *   hist      - 1D tensor containing the values of the histogram.
- *   bin_edges - 1D tensor containing the edges of the histogram bins. Contains hist.numel() + 1 elements.
+ *   hist      - D-dimensional tensor containing the values of the histogram.
+ *   bin_edges - tensor list of length D containing the edges of the histogram bins in each dimension.
  *               Bins include their left edge and exclude their right edge, with the exception of the
- *               rightmost bin which includes both of its edges.
+ *               rightmost bin in each dimension which includes both of its edges.
  *
  * Restrictions are defined in histogram_check_inputs() and in select_outer_bin_edges().
  */
 
 namespace at { namespace native {
 
-DEFINE_DISPATCH(histogram_stub);
-
-DEFINE_DISPATCH(histogram_linear_stub);
+DEFINE_DISPATCH(histogramdd_stub);
+DEFINE_DISPATCH(histogramdd_linear_stub);
 
 namespace {
 
 /* Checks properties of input tensors input, bins, and weight.
  */
-void histogram_check_inputs(const Tensor& input, const Tensor& bins, const c10::optional<Tensor>& weight) {
-    TORCH_CHECK(input.dtype() == bins.dtype(), "torch.histogram: input tensor and bins tensor should",
-            " have the same dtype, but got input ", input.dtype(), " and bins ", bins.dtype());
+void histogramdd_check_inputs(const Tensor& input, const TensorList& bins, const c10::optional<Tensor>& weight) {
+    TORCH_CHECK(input.dim() >= 2, "torch.histogramdd: input tensor should have at least 2 dimensions");
+
+    const int64_t D = input.size(-1);
+
+    TORCH_CHECK(int64_t(bins.size()) == D, "torch.histogramdd: expected ", D, " sequences of bin edges for a ", D,
+                "-dimensional histogram but got ", bins.size());
 
-    TORCH_CHECK(bins.dim() == 1, "torch.histogram: bins tensor should have dimension 1,",
-            " but got ", bins.dim(), " dimension");
+    for (int64_t dim = 0; dim < D; dim++) {
+        TORCH_CHECK(input.dtype() == bins[dim].dtype(), "torch.histogramdd: input tensor and bins tensor should",
```

Comment:
nit: the acquisition of input.dtype() can occur once outside the loop and the acquisition of bins[dim].dtype() can occur once at the start of each iteration of the loop